src/lib/helpers.js

```diff
@@ -0,0 +1,126 @@
+const { BigNumber, ethers } = require('ethers');
+const axios = require('axios');
+
+const config = require('../config');
+const {
+  TRANCHE_DURATION_DAYS,
+  MAX_ACTIVE_TRANCHES,
+  PRICE_CHANGE_PER_DAY,
+  NXM_PER_ALLOCATION_UNIT,
+  TARGET_PRICE_DENOMINATOR,
+  SURGE_THRESHOLD_RATIO,
+  SURGE_PRICE_RATIO,
+} = require('./constants');
+
+async function getContractFactory(provider) {
+  const url = config.get('contractsUrl');
+  const {
+    data: {
+      mainnet: { abis },
+    },
+  } = await axios.get(url);
+
+  const data = abis.reduce((acc, contract) => {
+    const { code, address, contractName, contractAbi } = contract;
+    acc[code] = {
+      address,
+      contractName,
+      abi: JSON.parse(contractAbi),
+    };
+    return acc;
+  }, {});
+
+  return async code => {
+    const { abi, address } = data[code];
+    return new ethers.Contract(address, abi, provider);
+  };
+}
+
+// offest in days
+function calculateTranche(offset = 0) {
+  return Math.floor((Date.now() / 86_400_000 + offset) / TRANCHE_DURATION_DAYS);
+}
```

Comment:
Please use timestamps in seconds everywhere. `TRANCHE_DURATION` in seconds, `offset` in seconds, etc. This function is likely to have some precision loss because you're operating on floats and it might error out when the timestamp is exact but you might get .999999